### DISCUSSION TITLE
로그아웃 시 캐싱된 쿼리를 어떻게 제거할 것인가?

### DISCUSSION THREAD

**Discussion by euijinkk** [woowacourse-teams/2022-kkogkkog#509] [Ideas]
### 현재 방식의 문제점
현재는 remove로 useFetchMe의 쿼리만 지워주고 있다. 그래서 아래처럼 다른 쿼리들이 남아있다. inactive한 쿼리들이 여전히 살아있고, 다른 계정으로 로그인하게 되면, 이전 쿼리 데이터가 조회될 수도 있다.
<img width="709" alt="KakaoTalk_Photo_2022-10-20-10-43-48" src="https://user-images.githubusercontent.com/24906022/196839830-33b35ad0-05f9-48ac-9aaf-32bf9c661730.png">

### 해결책 고민

1. 현재 존재하는 모든 캐시를 지워준다. (`queryClient.clear()`)
    - 안전하지만, 지워질 필요가 없는 query(인증 정보에 상관없이 동일한 정보)도 지워진다.
2. 제거가 필요한 query (인증 정보에 따라 다르게 보여지는 정보) 만을 따로 remove 해준다.
    - 인증 정보에 따라 다른 값을 반환하는 모든 query를 remove 해주어야 한다. 도메인이 더 커지고, remove 해주어야할 요소가 많아질때 하나하나 찾아서 적어주어야 한다. 새로운 query가 생길때마다 이것이 로그아웃 시에 제거되어야하는지 고민하고 추가해주어야 하는 과정이 필요하다. 변경이 발생할 때마다 바꿔줘야하는 부분이 많다는 점이 단점이다.
        
        아래와 같이 하면 도메인별로 간단하게 초기화시킬 수 있지만, 도메인 내에서 제거 필요 여부가 다른 Query들이 존재한다면 더욱 복잡해질 수 있다.
        
        ```jsx
        export const useInvalidateAllCouponQueries = () => {
          const queryClient = useQueryClient();
        
          Object.entries(QUERY_KEY).forEach(([k, v]) => {
            queryClient.invalidateQueries(v);
          });
        };
        ```
  
### 결론
- 인증 정보에 상관없는 정보가 많은 서비스이다 ⇒ 2번을 채택하여 하나하나 remove 해주는 작업을 고민할 수 있다. 그게 필요한 캐싱을 살려준다는 점에서 효율적일 것이다.
- 대부분의 정보가 인증 정보에 의존하고 있는 서비스이다 ⇒ 1번을 채택하여 모든 query를 clear 해준다. 필요한 소수의 캐싱을 초기화해버리는 단점을 받아들이면서, 불필요한 개발 요소를 줄인다.

### 제안
우리팀은 현재 인증정보에 상관없는 쿼리가 `useFetchUnregisteredCouponByCode` 밖에 없고, 이 쿼리의 캐싱을 유지함으로써 얻는 효용이 적으므로 모든 캐시를 제거하는 방향을 택한다. 
이후 필요에 따라 다른 방법을 도입해본다.

**Comment by juunzzi** [comment]
## 로그아웃 이슈에 대해 해결 방식 및 각 방식 별 제 생각을 남겨보았습니다

현재 방식의 문제점은 다음과 같습니다.

현재 방식 : logout 시 me 쿼리를 remove

문제점 : 만약 다른 계정으로 로그인하면 ? me 쿼리는 제거되나 인증 정보와 관련된 쿼리(쿠폰, 미등록 쿠폰 등)가 날라가지 않기에 다른 계정으로 로그인할 시 이전 계정의 정보를 확인하게 된다. (크나큰 문제라고 생각)

해결 방안 1 : queryClient.clear()
해결 방안 2 : 도메인 별로 각자의 쿼리를 모두 날려버리는 메소드를 작성하여 유저 정보와 연관되어 있는 도메인 쿼리만 날리도록 한다.

### 1. queryClient.clear()

장점 : 도메인 별로 함수를 작성하지 않아도 전역의 모든 캐시 데이터를 제거할 수 있다.

단점 : 인증 정보와 관련되지 않은 `remote data` 영역 역시 제거되기에 문제가 될 수 있다.
단점 : 언제나 기억할 수 있는게 아니기에 **추후 디버깅 과정, 유지보수 과정**이 어려워질 수 있다.

### 2. 도메인 별로 각자의 쿼리를 모두 날려버리는 메소드를 작성하여 유저 정보와 연관되어 있는 도메인 쿼리만 날리도록 한다.

장점 : 제거가 필요한 도메인만 제거할 수 있다. 인증 정보와 관련되지 않은 쿼리는 제거되지 않기에 이 후 **유지보수나 디버깅 과정**이 편해진다.

단점 : 도메인 영역이 생길 때마다 함수를 작성하는 노고가 들어간다.

## 서제션

확실히 2번 방식이 나아보입니다. 당장 시간은 조금 더  들어갈 지라도 필요한 함수를 만들어, 제거가 필요한 대상만 제거한다. 라는 아이디어는 이 후 개발 과정에 유리함을 줄 것이라 생각합니다.

하지만 1번 방식도 나쁘지 않습니다. 현재 우리 프로젝트 구조 상 인증 정보와 관련되지 않은 `remote data` 영역, 즉 도메인 영역이 생길 것이라 상상하기는 힘듭니다. 또한 현재 프로젝트 상황이 데모로 인해 긴급하기에 너무 미래를 생각해 2번 방식을 채택할 필요는 없다고 보여집니다.

둘 다 현재 프로젝트 구조 상 단점만을 갖지 않기에 어떤 방식을 채택하더라도 수용할 수 있습니다 !!
